File: src/interpreter/objects.rs
//! # Objects
//! 4/22/2026 - Nyx
//! 
//! An identical Zig implementation can be found at: ./objects.zig

/// Hashed Identifier
/// 
/// An identifier that was hashed into a smaller sequence to allow for more efficient memory management.
pub type IdentifierHash = usize;

pub enum Object {
    Byte(i8),
    UByte(u8),
    Short(i16),
    UShort(u16),
    Int(i32),
    UInt(u32),
    Long(i64),
    ULong(u64),

    Float(f32),
    Double(f64),

    Identifier(IdentifierHash),
}

pub enum ObjectRef {
    Byte(*mut i8),
    UByte(*mut u8),
    Short(*mut i16),
    UShort(*mut u16),
    Int(*mut i32),
    UInt(*mut u32),
    Long(*mut i64),
    ULong(*mut u64),

    Float(*mut f32),
    Double(*mut f64),

    Identifier(*mut IdentifierHash),
}

#[repr(C)]
pub struct RawObject {
    pub type_: ObjectType,
    pub raw: ObjectRaw,
}

#[repr(C)]
pub struct RawObjectRef {
    pub type_: ObjectType,
    pub raw: ObjectRawRef,
}

#[repr(u8)]
pub enum ObjectType {
    Byte,
    UByte,
    Short,
    UShort,
    Int,
    UInt,
    Long,
    ULong,

    Float,
    Double,

    Identifier,
}

#[repr(C)]
pub union ObjectRaw {
    byte: i8,
    ubyte: u8,
    short: i16,
    ushort: u16,
    int: i32,
    uint: u32,
    long: i64,
    ulong: u64,

    float: f32,
    double: f64,

    identifier: IdentifierHash,
}

#[repr(C)]
pub union ObjectRawRef {
    byte: *mut i8,
    ubyte: *mut u8,
    short: *mut i16,
    ushort: *mut u16,
    int: *mut i32,
    uint: *mut u32,
    long: *mut i64,
    ulong: *mut u64,

    float: *mut f32,
    double: *mut f64,

    identifier: *mut IdentifierHash,
}

impl Object {
    pub fn raw(self) -> RawObject {
        match self {
            Self::Byte(i) => return RawObject { type_: ObjectType::Byte, raw: ObjectRaw { byte: i}},
            Self::UByte(i) => return RawObject { type_: ObjectType::UByte, raw: ObjectRaw { ubyte: i } },
            Self::Short(i) => return RawObject { type_: ObjectType::Short, raw: ObjectRaw { short: i } },
            Self::UShort(i) => return RawObject { type_: ObjectType::UShort, raw: ObjectRaw { ushort: i } },
            Self::Int(i) => return RawObject { type_: ObjectType::Int, raw: ObjectRaw { int: i } },
            Self::UInt(i) => return RawObject { type_: ObjectType::UInt, raw: ObjectRaw { uint: i } },
            Self::Long(i) => return RawObject { type_: ObjectType::Long, raw: ObjectRaw { long: i } },
            Self::ULong(i) => return RawObject { type_: ObjectType::ULong, raw: ObjectRaw { ulong: i } },
            Self::Float(i) => return RawObject { type_: ObjectType::Float, raw: ObjectRaw { float: i } },
            Self::Double(i) => return RawObject { type_: ObjectType::Double, raw: ObjectRaw { double: i } },
            Self::Identifier(i) => return RawObject { type_: ObjectType::Identifier, raw: ObjectRaw { identifier: i } },
        }
    }
}

impl ObjectRef {
    pub fn raw(&self) -> RawObjectRef {
        match self {
            Self::Byte(i) => return RawObjectRef { type_: ObjectType::Byte, raw: ObjectRawRef { byte: *i }},
            Self::UByte(i) => return RawObjectRef { type_: ObjectType::UByte, raw: ObjectRawRef { ubyte: *i } },
            Self::Short(i) => return RawObjectRef { type_: ObjectType::Short, raw: ObjectRawRef { short: *i } },
            Self::UShort(i) => return RawObjectRef { type_: ObjectType::UShort, raw: ObjectRawRef { ushort: *i } },
            Self::Int(i) => return RawObjectRef { type_: ObjectType::Int, raw: ObjectRawRef { int: *i } },
            Self::UInt(i) => return RawObjectRef { type_: ObjectType::UInt, raw: ObjectRawRef { uint: *i } },
            Self::Long(i) => return RawObjectRef { type_: ObjectType::Long, raw: ObjectRawRef { long: *i } },
            Self::ULong(i) => return RawObjectRef { type_: ObjectType::ULong, raw: ObjectRawRef { ulong: *i } },
            Self::Float(i) => return RawObjectRef { type_: ObjectType::Float, raw: ObjectRawRef { float: *i } },
            Self::Double(i) => return RawObjectRef { type_: ObjectType::Double, raw: ObjectRawRef { double: *i } },
            Self::Identifier(i) => return RawObjectRef { type_: ObjectType::Identifier, raw: ObjectRawRef { identifier: *i } },
        }
    }
}

impl RawObject {
    pub fn pack(self) -> Object {
        unsafe {
            match self.type_ {
                ObjectType::Byte => Object::Byte(self.raw.byte),
                ObjectType::UByte => Object::UByte(self.raw.ubyte),
                ObjectType::Short => Object::Short(self.raw.short),
                ObjectType::UShort => Object::UShort(self.raw.ushort),
                ObjectType::Int => Object::Int(self.raw.int),
                ObjectType::UInt => Object::UInt(self.raw.uint),
                ObjectType::Long => Object::Long(self.raw.long),
                ObjectType::ULong => Object::ULong(self.raw.ulong),
                ObjectType::Float => Object::Float(self.raw.float),
                ObjectType::Double => Object::Double(self.raw.double),
                ObjectType::Identifier => Object::Identifier(self.raw.identifier),
            }
        }
    }
}

impl RawObjectRef {
    pub fn pack(&self) -> ObjectRef {
        unsafe {
            match self.type_ {
                ObjectType::Byte => ObjectRef::Byte(self.raw.byte),
                ObjectType::UByte => ObjectRef::UByte(self.raw.ubyte),
                ObjectType::Short => ObjectRef::Short(self.raw.short),
                ObjectType::UShort => ObjectRef::UShort(self.raw.ushort),
                ObjectType::Int => ObjectRef::Int(self.raw.int),
                ObjectType::UInt => ObjectRef::UInt(self.raw.uint),
                ObjectType::Long => ObjectRef::Long(self.raw.long),
                ObjectType::ULong => ObjectRef::ULong(self.raw.ulong),
                ObjectType::Float => ObjectRef::Float(self.raw.float),
                ObjectType::Double => ObjectRef::Double(self.raw.double),
                ObjectType::Identifier => ObjectRef::Identifier(self.raw.identifier),
            }
        }
    }
}

impl std::fmt::Display for Object {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Byte(i) => write!(f, "{}", i),
            Self::UByte(i) => write!(f, "{}", i),
            Self::Short(i) => write!(f, "{}", i),
            Self::UShort(i) => write!(f, "{}", i),
            Self::Int(i) => write!(f, "{}", i),
            Self::UInt(i) => write!(f, "{}", i),
            Self::Long(i) => write!(f, "{}", i),
            Self::ULong(i) => write!(f, "{}", i),
            Self::Float(i) => write!(f, "{}", i),
            Self::Double(i) => write!(f, "{}", i),
            Self::Identifier(i) => write!(f, "{}", i),
        }
    }
}


impl Into<Option<i8>> for Object {
    fn into(self) -> Option<i8> {
        match self {
            Self::Byte(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<u8>> for Object {
    fn into(self) -> Option<u8> {
        match self {
            Self::UByte(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<i16>> for Object {
    fn into(self) -> Option<i16> {
        match self {
            Self::Short(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<u16>> for Object {
    fn into(self) -> Option<u16> {
        match self {
            Self::UShort(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<i32>> for Object {
    fn into(self) -> Option<i32> {
        match self {
            Self::Int(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<u32>> for Object {
    fn into(self) -> Option<u32> {
        match self {
            Self::UInt(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<i64>> for Object {
    fn into(self) -> Option<i64> {
        match self {
            Self::Long(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<u64>> for Object {
    fn into(self) -> Option<u64> {
        match self {
            Self::ULong(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<f32>> for Object {
    fn into(self) -> Option<f32> {
        match self {
            Self::Float(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<f64>> for Object {
    fn into(self) -> Option<f64> {
        match self {
            Self::Double(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<IdentifierHash>> for Object {
    fn into(self) -> Option<IdentifierHash> {
        match self {
            Self::Identifier(i) => Some(i),
            _ => None,
        }
    }
}



impl Into<Option<i8>> for ObjectRef {
    fn into(self) -> Option<i8> {
        match self {
            Self::Byte(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<u8>> for ObjectRef {
    fn into(self) -> Option<u8> {
        match self {
            Self::UByte(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<i16>> for ObjectRef {
    fn into(self) -> Option<i16> {
        match self {
            Self::Short(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<u16>> for ObjectRef {
    fn into(self) -> Option<u16> {
        match self {
            Self::UShort(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<i32>> for ObjectRef {
    fn into(self) -> Option<i32> {
        match self {
            Self::Int(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<u32>> for ObjectRef {
    fn into(self) -> Option<u32> {
        match self {
            Self::UInt(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<i64>> for ObjectRef {
    fn into(self) -> Option<i64> {
        match self {
            Self::Long(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<u64>> for ObjectRef {
    fn into(self) -> Option<u64> {
        match self {
            Self::ULong(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<f32>> for ObjectRef {
    fn into(self) -> Option<f32> {
        match self {
            Self::Float(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<f64>> for ObjectRef {
    fn into(self) -> Option<f64> {
        match self {
            Self::Double(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<IdentifierHash>> for ObjectRef {
    fn into(self) -> Option<IdentifierHash> {
        match self {
            Self::Identifier(i) => unsafe { Some(*i) },
            _ => None,
        }
    }
}

impl Into<Option<*mut i8>> for ObjectRef {
    fn into(self) -> Option<*mut i8> {
        match self {
            Self::Byte(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut u8>> for ObjectRef {
    fn into(self) -> Option<*mut u8> {
        match self {
            Self::UByte(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut i16>> for ObjectRef {
    fn into(self) -> Option<*mut i16> {
        match self {
            Self::Short(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut u16>> for ObjectRef {
    fn into(self) -> Option<*mut u16> {
        match self {
            Self::UShort(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut i32>> for ObjectRef {
    fn into(self) -> Option<*mut i32> {
        match self {
            Self::Int(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut u32>> for ObjectRef {
    fn into(self) -> Option<*mut u32> {
        match self {
            Self::UInt(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut i64>> for ObjectRef {
    fn into(self) -> Option<*mut i64> {
        match self {
            Self::Long(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut u64>> for ObjectRef {
    fn into(self) -> Option<*mut u64> {
        match self {
            Self::ULong(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut f32>> for ObjectRef {
    fn into(self) -> Option<*mut f32> {
        match self {
            Self::Float(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut f64>> for ObjectRef {
    fn into(self) -> Option<*mut f64> {
        match self {
            Self::Double(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<*mut IdentifierHash>> for ObjectRef {
    fn into(self) -> Option<*mut IdentifierHash> {
        match self {
            Self::Identifier(i) => Some(i),
            _ => None,
        }
    }
}

impl Into<Option<&mut i8>> for ObjectRef {
    fn into(self) -> Option<&'static mut i8> {
        match self {
            Self::Byte(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut u8>> for ObjectRef {
    fn into(self) -> Option<&'static mut u8> {
        match self {
            Self::UByte(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut i16>> for ObjectRef {
    fn into(self) -> Option<&'static mut i16> {
        match self {
            Self::Short(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut u16>> for ObjectRef {
    fn into(self) -> Option<&'static mut u16> {
        match self {
            Self::UShort(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut i32>> for ObjectRef {
    fn into(self) -> Option<&'static mut i32> {
        match self {
            Self::Int(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut u32>> for ObjectRef {
    fn into(self) -> Option<&'static mut u32> {
        match self {
            Self::UInt(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut i64>> for ObjectRef {
    fn into(self) -> Option<&'static mut i64> {
        match self {
            Self::Long(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut u64>> for ObjectRef {
    fn into(self) -> Option<&'static mut u64> {
        match self {
            Self::ULong(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut f32>> for ObjectRef {
    fn into(self) -> Option<&'static mut f32> {
        match self {
            Self::Float(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut f64>> for ObjectRef {
    fn into(self) -> Option<&'static mut f64> {
        match self {
            Self::Double(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

impl Into<Option<&mut IdentifierHash>> for ObjectRef {
    fn into(self) -> Option<&'static mut IdentifierHash> {
        match self {
            Self::Identifier(i) => unsafe { i.as_mut() },
            _ => None,
        }
    }
}

File: src/interpreter-old/objects.zig
//! # Objects
//! 4/22/2026 - Nyx
//! 
//! An identical Rust implementation can be found at: ./objects.rs

// INCLUDES -----
const std = @import("std");

/// Hashed Identifier
/// 
/// An identifier that was hashed into a smaller sequence to allow for more efficient memory management.
pub const IdentifierHash = usize;

/// Object
/// 
/// An ambiguous object stored by the interpreter.
/// Objects are set up in a special way that lets them be converted 
/// into a ffi compatible raw type and passed between Zig and Rust.
pub const Object = union(Type) {
    byte: i8,
    ubyte: u8,
    short: i16,
    ushort: u16,
    int: i32,
    uint: u32,
    long: i64,
    ulong: u64,

    float: f32,
    double: f64,

    identifier: IdentifierHash,

    pub const Type = enum(u8) {
        byte,
        ubyte,
        short,
        ushort,
        int,
        uint,
        long,
        ulong,

        float,
        double,

        identifier,
    };

    /// Raw extern union of the data
    pub const Raw = extern union {
        byte: i8,
        ubyte: u8,
        short: i16,
        ushort: u16,
        int: i32,
        uint: u32,
        long: i64,
        ulong: u64,

        float: f32,
        double: f64,

        identifier: IdentifierHash,
    };

    /// Raw version of the object packed identically but ffi compatible.
    pub const RawObject = extern struct {
        type: Type,
        raw: Raw,

        pub fn pack(self: RawObject) Object {
            switch (self.type) {
                .byte => return .{ .byte = self.raw.byte },
                .ubyte => return .{ .ubyte = self.raw.ubyte },
                .short => return .{ .short = self.raw.short },
                .ushort => return .{ .ushort = self.raw.ushort },
                .int => return .{ .int = self.raw.int },
                .uint => return .{ .uint = self.raw.uint },
                .long => return .{ .long = self.raw.long },
                .ulong => return .{ .ulong = self.raw.ulong },
                .float => return .{ .float = self.raw.float },
                .double => return .{ .double = self.raw.double },
                .identifier => return .{ .identifier = self.raw.identifier },
            }
        }
    };

    pub const ObjectRef = union(Type) {
        byte: *i8,
        ubyte: *u8,
        short: *i16,
        ushort: *u16,
        int: *i32,
        uint: *u32,
        long: *i64,
        ulong: *u64,

        float: *f32,
        double: *f64,

        identifier: *IdentifierHash,

        pub const RawRef = extern union {
            byte: *i8,
            ubyte: *u8,
            short: *i16,
            ushort: *u16,
            int: *i32,
            uint: *u32,
            long: *i64,
            ulong: *u64,

            float: *f32,
            double: *f64,

            identifier: *IdentifierHash,
        };

        pub const RawObjectRef = extern struct {
            type: Type,
            raw: RawRef,

            pub fn pack(self: RawObjectRef) ObjectRef {
                switch (self.type) {
                    .byte => return .{ .byte = self.raw.byte },
                    .ubyte => return .{ .ubyte = self.raw.ubyte },
                    .short => return .{ .short = self.raw.short },
                    .ushort => return .{ .ushort = self.raw.ushort },
                    .int => return .{ .int = self.raw.int },
                    .uint => return .{ .uint = self.raw.uint },
                    .long => return .{ .long = self.raw.long },
                    .ulong => return .{ .ulong = self.raw.ulong },
                    .float => return .{ .float = self.raw.float },
                    .double => return .{ .double = self.raw.double },
                    .identifier => return .{ .identifier = self.raw.identifier },
                }
            }
        };

        pub fn raw(self: ObjectRef) RawObjectRef {
            switch (self) {
            .byte => |i| return .{ .type = .byte, .raw = .{ .byte = i } },
            .ubyte => |i| return .{ .type = .ubyte, .raw = .{ .ubyte = i } },
            .short => |i| return .{ .type = .short, .raw = .{ .short = i } },
            .ushort => |i| return .{ .type = .ushort, .raw = .{ .ushort = i } },
            .int => |i| return .{ .type = .int, .raw = .{ .int = i } },
            .uint => |i| return .{ .type = .uint, .raw = .{ .uint = i } },
            .long => |i| return .{ .type = .long, .raw = .{ .long = i } },
            .ulong => |i| return .{ .type = .ulong, .raw = .{ .ulong = i } },
            .float => |i| return .{ .type = .float, .raw = .{ .float = i } },
            .double => |i| return .{ .type = .double, .raw = .{ .double = i } },
            .identifier => |i| return .{ .type = .identifier, .raw = .{ .identifier = i } },
        }
        }
    };

    pub fn raw(self: Object) RawObject {
        switch (self) {
            .byte => |i| return .{ .type = .byte, .raw = .{ .byte = i } },
            .ubyte => |i| return .{ .type = .ubyte, .raw = .{ .ubyte = i } },
            .short => |i| return .{ .type = .short, .raw = .{ .short = i } },
            .ushort => |i| return .{ .type = .ushort, .raw = .{ .ushort = i } },
            .int => |i| return .{ .type = .int, .raw = .{ .int = i } },
            .uint => |i| return .{ .type = .uint, .raw = .{ .uint = i } },
            .long => |i| return .{ .type = .long, .raw = .{ .long = i } },
            .ulong => |i| return .{ .type = .ulong, .raw = .{ .ulong = i } },
            .float => |i| return .{ .type = .float, .raw = .{ .float = i } },
            .double => |i| return .{ .type = .double, .raw = .{ .double = i } },
            .identifier => |i| return .{ .type = .identifier, .raw = .{ .identifier = i } },
        }
    }

    pub fn ref(self: *Object) ObjectRef {
        switch (self.*) {
            .byte => return .{ .byte = &self.byte },
            .ubyte => return .{ .ubyte = &self.ubyte },
            .short => return .{ .short = &self.short },
            .ushort => return .{ .ushort = &self.ushort },
            .int => return .{ .int = &self.int },
            .uint => return .{ .uint = &self.uint },
            .long => return .{ .long = &self.long },
            .ulong => return .{ .ulong = &self.ulong },
            .float => return .{ .float = &self.float },
            .double => return .{ .double = &self.double },
            .identifier => return .{ .identifier = &self.identifier },
        }
    }

    pub fn format(self: *const Object, writer: *std.io.Writer) std.Io.Writer.Error!void {
        _ = self;
        try writer.print("LIT", .{});
    }

};